### PR TITLE
(cloud-merge) Support shadow tablet to do cumulative compaction in cloud mode

### PR DESCRIPTION
**Pull request by Lchangliang**
In cloud mode, when do schema change, shadow tablet encounters -235 because it cant do cumulative compaction in the case of a large number of loads. And it will prevents the user from continuing to loads.
Implementation details:
1. When start schema change, record the end convert rowset version `alter_version` into SchemaChangeJob.
2. For origin tablet, only can do base compaction in [0, `alter_version`] and do cumulative compaction in (`alter_version`, N]. can not do compaction across `alter_verison` such as compaction [a, `alter_version` + n].
3. For shadow tablet, cannot do base compaction and and do cumulative compaction in (`alter_version`, N].
4. When the schema change failed because FE or BE coredump, it will retry. When retry the schema change, it will get the `alter_version` from meta_serive, and continue to do it.
5. When finish the schema change job or cancel it, we need to clear the schema change job. Before this pr, it will cover by next schema change.

### CODE REVIEW

**Comment by Lchangliang** [comment]
run buildall

**Comment by dataroaring** [comment]
run buildall

**Comment by Lchangliang** [comment]
run buildall

File: be/src/cloud/cloud_tablet.h [hunk 1]
```diff
@@ -145,6 +145,9 @@ class CloudTablet final : public BaseTablet {
         _last_base_compaction_schedule_millis = millis;
     }
 
+    int64_t alter_version() const { return _alter_version; }
+    void set_alter_version(int64_t alter_version) { _alter_version = alter_version; }
```
Comment (luwei16) [inline be/src/cloud/cloud_tablet.h:149]: When to reset alter version = -1

File: be/src/cloud/cloud_cumulative_compaction.cpp [hunk 2]
```diff
@@ -48,7 +48,8 @@ CloudCumulativeCompaction::CloudCumulativeCompaction(CloudStorageEngine& engine,
 CloudCumulativeCompaction::~CloudCumulativeCompaction() = default;
 
 Status CloudCumulativeCompaction::prepare_compact() {
-    if (_tablet->tablet_state() != TABLET_RUNNING) {
+    if (_tablet->tablet_state() != TABLET_RUNNING &&
+        dynamic_cast<CloudTablet*>(_tablet.get())->alter_version() == -1) {
```
Comment (luwei16) [inline be/src/cloud/cloud_cumulative_compaction.cpp:52]: While waiting for the waterline, the new tablet still cannot do compaction?

File: be/src/cloud/cloud_tablet.h (hunk 1 above)
Comment (Lchangliang) [reply be/src/cloud/cloud_tablet.h:149]: BE restart or quit from process_alter_tablet. In process_alter_tablet, there have a defer to change it.

File: be/src/cloud/cloud_cumulative_compaction.cpp (hunk 2 above)
Comment (Lchangliang) [reply be/src/cloud/cloud_cumulative_compaction.cpp:52]: _tablet->tablet_state() is always not equals TABLET_RUNNING until the schema change done. So this condition is not concerned with schema change job intermediate state.

File: gensrc/proto/cloud.proto [hunk 3]
```diff
@@ -536,6 +536,7 @@ message TabletCompactionJobPB {
     optional int64 lease = 23; // prepare
     optional int64 delete_bitmap_lock_initiator = 24;
     optional int64 full_compaction_cnt = 25;  // prepare
+    optional bool judge_input_versions_range = 26;
```
Comment (gavinchou) [inline gensrc/proto/cloud.proto:539]: judge_input_versions_range -> check_input_versions_range

File: fe/fe-core/src/main/java/org/apache/doris/cloud/datasource/CloudInternalCatalog.java [hunk 4]
```diff
@@ -810,6 +810,48 @@ private void dropCloudPartition(long dbId, long tableId, List<Long> partitionIds
...
+                LOG.warn("tryTimes:{}, dropIndex RpcException", tryTimes, e);
+                if (tryTimes + 1 >= Config.metaServiceRpcRetryTimes()) {
+                    throw new DdlException(e.getMessage());
+                }
+            }
+            sleepSeveralMs();
+        }
+
+        if (response.getStatus().getCode() != Cloud.MetaServiceCode.OK) {
+            LOG.warn("dropIndex response: {} ", response);
```
Comment (gavinchou) [inline fe/fe-core/src/main/java/org/apache/doris/cloud/datasource/CloudInternalCatalog.java:851]: typo? wrong log content

File: fe/fe-core/src/main/java/org/apache/doris/cloud/datasource/CloudInternalCatalog.java [hunk 5]
```diff
@@ -810,6 +810,48 @@ private void dropCloudPartition(long dbId, long tableId, List<Long> partitionIds
...
+        Cloud.FinishTabletJobResponse response = null;
+        int tryTimes = 0;
+        while (tryTimes++ < Config.metaServiceRpcRetryTimes()) {
+            try {
+                response = MetaServiceProxy.getInstance().finishTabletJob(request);
+                if (response.getStatus().getCode() != Cloud.MetaServiceCode.KV_TXN_CONFLICT) {
+                    break;
+                }
+            } catch (RpcException e) {
+                LOG.warn("tryTimes:{}, dropIndex RpcException", tryTimes, e);
```
Comment (gavinchou) [inline fe/fe-core/src/main/java/org/apache/doris/cloud/datasource/CloudInternalCatalog.java:842]: typo?

File: fe/fe-core/src/main/java/org/apache/doris/alter/CloudSchemaChangeJobV2.java [hunk 6]
```diff
@@ -106,14 +110,39 @@ protected void commitShadowIndex() throws AlterCancelException {
...
+                    Long partitionId = data.getRowKey();
+                    Long shadowIndexId = data.getColumnKey();
+                    Long originIndexId = indexIdMap.get(shadowIndexId);
+                    Map<Long, Long> shadowTabletIdToOriginTabletId = data.getValue();
+                    for (Map.Entry<Long, Long> entry : shadowTabletIdToOriginTabletId.entrySet()) {
+                        Long originTabletId = entry.getValue();
+                        ((CloudInternalCatalog) Env.getCurrentInternalCatalog())
+                            .removeSchemaChangeJob(dbId, tableId, originIndexId, partitionId, originTabletId);
+                    }
+                }
```
Comment (gavinchou) [inline fe/fe-core/src/main/java/org/apache/doris/alter/CloudSchemaChangeJobV2.java:138]: log here:
* which table/index has been processed (id name etc.)
* how many talbets have been processed here.

File: fe/fe-core/src/main/java/org/apache/doris/alter/CloudSchemaChangeJobV2.java [hunk 7]
```diff
@@ -106,14 +110,39 @@ protected void commitShadowIndex() throws AlterCancelException {
...
+                break;
+            } catch (Exception e) {
+                LOG.warn("tryTimes:{}, onCancel exception:", tryTimes, e);
+            }
+            sleepSeveralSeconds();
+            tryTimes++;
```
Comment (gavinchou) [inline fe/fe-core/src/main/java/org/apache/doris/alter/CloudSchemaChangeJobV2.java:144]: What if it tries to abort the same tablet job multiple times?

File: fe/fe-core/src/main/java/org/apache/doris/alter/CloudRollupJobV2.java [hunk 8]
```diff
@@ -124,6 +124,15 @@ protected void onCancel() {
...
                     .dropMaterializedIndex(tableId, rollupIndexList, false);
+                for (Map.Entry<Long, Map<Long, Long>> partitionEntry : partitionIdToBaseRollupTabletIdMap.entrySet()) {
+                    Long partitionId = partitionEntry.getKey();
+                    Map<Long, Long> rollupTabletIdToBaseTabletId = partitionEntry.getValue();
+                    for (Map.Entry<Long, Long> tabletEntry : rollupTabletIdToBaseTabletId.entrySet()) {
+                        Long baseTabletId = tabletEntry.getValue();
+                        ((CloudInternalCatalog) Env.getCurrentInternalCatalog())
+                            .removeSchemaChangeJob(dbId, tableId, baseIndexId, partitionId, baseTabletId);
+                    }
+                }
```
Comment (gavinchou) [inline fe/fe-core/src/main/java/org/apache/doris/alter/CloudRollupJobV2.java:135]: log here:
* which table/index has been processed (id name etc.)
* how many talbets have been processed here.

File: cloud/src/meta-service/meta_service_job.cpp [hunk 9]
```diff
@@ -124,6 +136,17 @@ void start_compaction_job(MetaServiceCode& code, std::string& msg, std::stringst
     }
     while (err == TxnErrorCode::TXN_OK) {
         job_pb.ParseFromString(job_val);
+        if (job_pb.has_schema_change() && !check_compaction_input_verions(compaction, job_pb)) {
```
Comment (gavinchou) [inline cloud/src/meta-service/meta_service_job.cpp:139]: why not put `job_pb.has_schema_change()` into `check_compaction_input_verions()`

File: cloud/src/meta-service/meta_service_job.cpp [hunk 10]
```diff
@@ -553,30 +590,23 @@ void process_compaction_job(MetaServiceCode& code, std::string& msg, std::string
...
-        txn->put(job_key, job_val);
-        INSTANCE_LOG(INFO) << "lease tablet compaction job, tablet_id=" << tablet_id
-                           << " key=" << hex(job_key);
-        need_commit = true;
-        return;
+    bool abort_compaction = false;
+    if (recorded_job.has_schema_change() && request->action() == FinishTabletJobRequest::COMMIT &&
+        !check_compaction_input_verions(compaction, recorded_job)) {
+        SS << "Check compaction input versions failed in schema change. input_version_start="
+           << compaction.input_versions(0) << " input_version_end=" << compaction.input_versions(1)
```
Comment (gavinchou) [inline cloud/src/meta-service/meta_service_job.cpp:597]: `SS` typo? it should be lowercase

File: cloud/src/meta-service/meta_service_job.cpp [hunk 11]
```diff
@@ -49,6 +49,18 @@ namespace doris::cloud {
 static constexpr int COMPACTION_DELETE_BITMAP_LOCK_ID = -1;
 static constexpr int SCHEMA_CHANGE_DELETE_BITMAP_LOCK_ID = -2;
 
+bool check_compaction_input_verions(const TabletCompactionJobPB& compaction,
```
Comment (gavinchou) [inline cloud/src/meta-service/meta_service_job.cpp:52]: Add sufficient comment what it checks how it works

File: be/src/cloud/cloud_base_compaction.cpp [hunk 12]
```diff
@@ -97,6 +99,18 @@ Status CloudBaseCompaction::prepare_compact() {
         } else if (resp.status().code() == cloud::TABLET_NOT_FOUND) {
             // tablet not found
             cloud_tablet()->clear_cache();
+        } else if (resp.status().code() == cloud::JOB_CHECK_ALTER_VERSION_FAIL) {
+            (dynamic_cast<CloudTablet*>(_tablet.get()))->set_alter_version(resp.alter_version());
```
Comment (gavinchou) [inline be/src/cloud/cloud_base_compaction.cpp:103]: just static_cast if you dont check the cast result.

File: be/src/cloud/cloud_base_compaction.cpp [hunk 13]
```diff
@@ -314,6 +328,18 @@ Status CloudBaseCompaction::modify_rowsets() {
     if (!st.ok()) {
         if (resp.status().code() == cloud::TABLET_NOT_FOUND) {
             cloud_tablet()->clear_cache();
+        } else if (resp.status().code() == cloud::JOB_CHECK_ALTER_VERSION_FAIL) {
+            (dynamic_cast<CloudTablet*>(_tablet.get()))->set_alter_version(resp.alter_version());
```
Comment (gavinchou) [inline be/src/cloud/cloud_base_compaction.cpp:332]: ditto

File: be/src/olap/schema_change.cpp [hunk 14]
```diff
@@ -23,6 +23,7 @@
 #include <memory>
 #include <mutex>
 #include <roaring/roaring.hh>
+#include <thread>
```
Comment (gavinchou) [inline be/src/olap/schema_change.cpp:26]: do not touch what you dont modify

File: be/src/cloud/cloud_schema_change_job.cpp [hunk 15]
```diff
@@ -86,13 +85,54 @@ Status CloudSchemaChangeJob::process_alter_tablet(const TAlterTabletReqV2& reque
...
+    auto* new_tablet_idx = sc_job->mutable_new_tablet_idx();
+    new_tablet_idx->set_tablet_id(_new_tablet->tablet_id());
+    new_tablet_idx->set_table_id(_new_tablet->table_id());
+    new_tablet_idx->set_index_id(_new_tablet->index_id());
+    new_tablet_idx->set_partition_id(_new_tablet->partition_id());
+    cloud::StartTabletJobResponse start_resp;
+    auto st = _cloud_storage_engine.meta_mgr().prepare_tablet_job(job, &start_resp);
+    if (!st.ok()) {
+        if (start_resp.status().code() == cloud::JOB_ALREADY_SUCCESS) {
+            st = _new_tablet->sync_rowsets();
```
Comment (gavinchou) [inline be/src/cloud/cloud_schema_change_job.cpp:110]: why do we need to sync rowsets explicitly?

File: cloud/test/meta_service_job_test.cpp [hunk 16]
```diff
@@ -58,6 +58,7 @@ void start_compaction_job(MetaService* meta_service, int64_t tablet_id, const st
     if (input_version.first > 0 && input_version.second > 0) {
         compaction->add_input_versions(input_version.first);
         compaction->add_input_versions(input_version.second);
+        compaction->set_judge_input_versions_range(true);
```
Comment (gavinchou) [inline cloud/test/meta_service_job_test.cpp:61]: what if the old version of BEs?

File: cloud/test/meta_service_job_test.cpp
Comment (gavinchou) [inline cloud/test/meta_service_job_test.cpp:1]: Add UT for the newly added code in meta_service_job.cpp

File: gensrc/proto/cloud.proto [hunk 17]
```diff
@@ -1324,6 +1327,7 @@ enum MetaServiceCode {
     JOB_ALREADY_SUCCESS = 5002;
     ROUTINE_LOAD_DATA_INCONSISTENT = 5003;
     ROUTINE_LOAD_PROGRESS_NOT_FOUND = 5004;
+    JOB_CHECK_ALTER_VERSION_FAIL = 5005;
```
Comment (gavinchou) [inline gensrc/proto/cloud.proto:1330]: what does "fail" mean? unmatched?

File: cloud/src/meta-service/meta_service_job.cpp (hunk 10 above)
Comment (Lchangliang) [reply cloud/src/meta-service/meta_service_job.cpp:597]: #define SS (ss << &__FILE__[get_file_name_offset(__FILE__)] << ":" << __LINE__ << " ")

File: be/src/cloud/cloud_schema_change_job.cpp (hunk 15 above)
Comment (Lchangliang) [reply be/src/cloud/cloud_schema_change_job.cpp:110]: The code used to be there. I don't think it's important, but it's best not to modify it.

File: gensrc/proto/cloud.proto (hunk 17 above)
Comment (Lchangliang) [reply gensrc/proto/cloud.proto:1330]: Does not comply with the rules of 
```
(compaction.type() == TabletCompactionJobPB_CompactionType_BASE &&
            compaction.input_versions(1) <= alter_version) ||
           (compaction.type() == TabletCompactionJobPB_CompactionType_CUMULATIVE &&
            compaction.input_versions(0) > alter_version)
```

**Comment by Lchangliang** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/alter/CloudSchemaChangeJobV2.java (hunk 7 above)
Comment (Lchangliang) [reply fe/fe-core/src/main/java/org/apache/doris/alter/CloudSchemaChangeJobV2.java:149]: Ms will check the message <origin_idx, shadow_idx, origin_tablet, shadow_tablet>. If it's not match, skip the request.

File: fe/fe-core/src/main/java/org/apache/doris/alter/CloudRollupJobV2.java (hunk 8 above)
Comment (Lchangliang) [reply fe/fe-core/src/main/java/org/apache/doris/alter/CloudRollupJobV2.java:140]: Ms will check the message <origin_idx, rollup_idx, origin_tablet, rollup_tablet>. If it's not match, skip the request.

File: cloud/test/meta_service_job_test.cpp (hunk 16 above)
Comment (Lchangliang) [reply cloud/test/meta_service_job_test.cpp]: The parameter is used to replace confi `enable_parallel_cumu_compaction`.
In old version of BEs, only if enable_parallel_cumu_compaction is true, input_versions will be set.
In new version of Bes, input_versions always are set, and use other param to replace `enable_parallel_cumu_compaction`.
Compatibility is also handled here.

File: be/src/cloud/cloud_meta_mgr.cpp [hunk 18]
```diff
@@ -448,7 +448,8 @@ Status CloudMetaMgr::sync_tablet_rowsets(CloudTablet* tablet, bool warmup_delta_
         int64_t now = duration_cast<seconds>(system_clock::now().time_since_epoch()).count();
         tablet->last_sync_time_s = now;
 
-        if (tablet->enable_unique_key_merge_on_write()) {
+        if (tablet->enable_unique_key_merge_on_write() &&
+            tablet->tablet_state() == TABLET_RUNNING) {
```
[the remark this answers is not in the document]
Comment (gavinchou) [reply be/src/cloud/cloud_meta_mgr.cpp:452]: add comment

File: be/src/cloud/cloud_cumulative_compaction.cpp [hunk 19]
```diff
@@ -48,7 +48,8 @@ CloudCumulativeCompaction::CloudCumulativeCompaction(CloudStorageEngine& engine,
 CloudCumulativeCompaction::~CloudCumulativeCompaction() = default;
 
 Status CloudCumulativeCompaction::prepare_compact() {
-    if (_tablet->tablet_state() != TABLET_RUNNING) {
+    if (_tablet->tablet_state() != TABLET_RUNNING && config::enable_new_tablet_do_compaction &&
```
Comment (gavinchou) [inline be/src/cloud/cloud_cumulative_compaction.cpp:51]: !config::enable_new_tablet_do_compaction

File: cloud/src/meta-service/meta_service_job.cpp [hunk 20]
```diff
@@ -49,6 +49,29 @@ namespace doris::cloud {
...
+// we will check they in prepare compaction and commit compaction.
+// 1. When if base compaction, we need to guarantee the end version
+// is less than or equal to alter_version.
+// 2. When if cu compaction, we need to guarantee the start version
+// is large than alter_version.
+bool check_compaction_input_verions(const TabletCompactionJobPB& compaction,
+                                    const TabletJobInfoPB& job_pb) {
+    if (!job_pb.has_schema_change() || !job_pb.schema_change().has_alter_version()) return true;
+    // compaction need to know [start_version, end_version]
+    DCHECK_EQ(compaction.input_versions_size(), 2) << proto_to_json(compaction);
```
Comment (gavinchou) [inline cloud/src/meta-service/meta_service_job.cpp:64]: if () logging

File: cloud/src/meta-service/meta_service_job.cpp [hunk 21]
```diff
@@ -290,6 +327,16 @@ void start_schema_change_job(MetaServiceCode& code, std::string& msg, std::strin
     std::string job_val;
     TabletJobInfoPB job_pb;
     err = txn->get(job_key, &job_val);
+    if (err == TxnErrorCode::TXN_OK) {
```
Comment (gavinchou) [inline cloud/src/meta-service/meta_service_job.cpp:330]: move to line 352

File: fe/fe-core/src/main/java/org/apache/doris/alter/CloudRollupJobV2.java [hunk 22]
```diff
@@ -124,6 +124,20 @@ protected void onCancel() {
...
                     .dropMaterializedIndex(tableId, rollupIndexList, false);
+                for (Map.Entry<Long, Map<Long, Long>> partitionEntry : partitionIdToBaseRollupTabletIdMap.entrySet()) {
+                    Long partitionId = partitionEntry.getKey();
+                    Map<Long, Long> rollupTabletIdToBaseTabletId = partitionEntry.getValue();
+                    for (Map.Entry<Long, Long> tabletEntry : rollupTabletIdToBaseTabletId.entrySet()) {
+                        Long rollupTabletId = tabletEntry.getKey();
+                        Long baseTabletId = tabletEntry.getValue();
+                        ((CloudInternalCatalog) Env.getCurrentInternalCatalog())
+                            .removeSchemaChangeJob(dbId, tableId, baseIndexId, rollupIndexId,
+                                    partitionId, baseTabletId, rollupTabletId);
```
Comment (gavinchou) [inline fe/fe-core/src/main/java/org/apache/doris/alter/CloudRollupJobV2.java:135]: try best to cancel sc job even there is a RPC error

File: fe/fe-core/src/main/java/org/apache/doris/alter/CloudSchemaChangeJobV2.java [hunk 23]
```diff
@@ -106,14 +110,44 @@ protected void commitShadowIndex() throws AlterCancelException {
...
+            try {
+                Set<Table.Cell<Long, Long, Map<Long, Long>>> tableSet = partitionIndexTabletMap.cellSet();
+                Iterator<Cell<Long, Long, Map<Long, Long>>> it = tableSet.iterator();
+                while (it.hasNext()) {
+                    Table.Cell<Long, Long, Map<Long, Long>> data = it.next();
+                    Long partitionId = data.getRowKey();
+                    Long shadowIndexId = data.getColumnKey();
+                    Long originIndexId = indexIdMap.get(shadowIndexId);
+                    Map<Long, Long> shadowTabletIdToOriginTabletId = data.getValue();
+                    for (Map.Entry<Long, Long> entry : shadowTabletIdToOriginTabletId.entrySet()) {
```
Comment (gavinchou) [inline fe/fe-core/src/main/java/org/apache/doris/alter/CloudSchemaChangeJobV2.java:133]: try best to cancel sc job even there is a RPC error

File: cloud/src/meta-service/meta_service_job.cpp [hunk 24]
```diff
@@ -961,11 +1027,48 @@ void process_schema_change_job(MetaServiceCode& code, std::string& msg, std::str
...
     //==========================================================================
     //                               Abort
     //==========================================================================
     if (request->action() == FinishTabletJobRequest::ABORT) {
-        // TODO(cyx)
+        if (schema_change.new_tablet_idx().index_id() ==
+                    recorded_schema_change.new_tablet_idx().index_id() &&
+            schema_change.new_tablet_idx().tablet_id() ==
+                    recorded_schema_change.new_tablet_idx().tablet_id()) {
+            // TODO(cyx)
```
Comment (gavinchou) [inline cloud/src/meta-service/meta_service_job.cpp:1059]: remove it

File: fe/fe-core/src/main/java/org/apache/doris/alter/CloudSchemaChangeJobV2.java (hunk 23 above)
Comment (Lchangliang) [reply fe/fe-core/src/main/java/org/apache/doris/alter/CloudSchemaChangeJobV2.java:133]:
```
// fe/fe-core/src/main/java/org/apache/doris/cloud/datasource/CloudInternalCatalog.java:removeSchemaChangeJob
int tryTimes = 0;
        while (tryTimes++ < Config.metaServiceRpcRetryTimes()) {
            try {
                response = MetaServiceProxy.getInstance().finishTabletJob(request);
                if (response.getStatus().getCode() != Cloud.MetaServiceCode.KV_TXN_CONFLICT) {
                    break;
                }
            } catch (RpcException e) {
                LOG.warn("tryTimes:{}, finishTabletJob RpcException", tryTimes, e);
                if (tryTimes + 1 >= Config.metaServiceRpcRetryTimes()) {
                    throw new DdlException(e.getMessage());
                }
            }
            sleepSeveralMs();
        }
```

**Comment by Lchangliang** [comment]
run buildall

**Comment by Lchangliang** [comment]
run buildall

**Comment by Lchangliang** [comment]
run buildall

**Comment by Lchangliang** [comment]
run buildall

**Comment by Lchangliang** [comment]
run cloud ut

**Comment by Lchangliang** [comment]
run buildall

**Comment by Lchangliang** [comment]
run buildall

**Comment by Lchangliang** [comment]
run p0

**Review by luwei16 — APPROVED** [review]

**Review by gavinchou — APPROVED** [review]

**Pull request merged**